client/client-lib/src/api.rs

```diff
@@ -357,30 +347,31 @@ impl<C: JsonRpcClient> WsFederationApi<C> {
         &self,
         method: &str,
         param: P,
+        strategy: &mut impl QueryStrategy<R>,
     ) -> Result<R> {
         let params = [serde_json::to_value(param).expect("encoding error")];
-        let requests = self
-            .members
-            .iter()
-            .map(|member| Box::pin(member.request(method, &params)));
-
-        let mut error = None;
-        let mut successes = 0;
-        for request in requests {
-            match request.await {
-                Ok(res) => {
-                    successes += 1;
-                    if successes == 2 * self.max_evil + 1 {
-                        return Ok(res);
-                    }
-                }
-                Err(e) => error = Some(e),
-            };
+        let mut futures = FuturesUnordered::new();
+
+        for member in &self.members {
+            futures.push(member.request(method, &params));
         }
 
-        Err(ApiError::RpcError(
-            error.expect("If there was no success there has to be an error"),
-        ))
+        loop {
+            match futures.next().await {
+                Some(result) => match strategy.process(result, self.max_evil) {
+                    QueryStep::Request(peers) => {
+                        for member in &self.members {
+                            if peers.contains(&member.peer_id) {
```

Comment:
If this is false does it mean that the guardian lied? Should we log that fact?

Reply:
So the strategy itself resolves bad guardians, the `QueryStep:Request` is how the strategy communicates to this `request` function that it wants to resend requests to some of the peers.

Reply:
My point is mostly about being able to warn the user that certain guardian is misbehaving. I guess we can't ignore for now, but I imagine we will eventually want some handling for it.